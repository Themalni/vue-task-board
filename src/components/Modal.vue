<template>
  <transition name="fade">
    <div class="overlay">
      <div class="modal" :class="{ 'w-25': small }">
        <div class="modal-header">
          <h3 class="modal-header-text text-white">{{ title }}</h3>
          <h3 v-if="status" class="modal-header-text text-white ml-1">({{ status }})</h3>
        </div>
        <div class="modal-body">
          <slot name="body"></slot>
        </div>
        <div class="modal-footer">
          <slot name="footer"></slot>
        </div>
      </div>
    </div>
  </transition>
</template>

<script>

export default {
  name: "Modal",
  props: {
    title: {
      type: String
    },
    status: {
      type: [String, Boolean],
      default: false
    },
    small: {
      type: Boolean,
      default: false
    }
  }
};
</script>

<style lang="scss">
.fade-enter-active,
.fade-leave-active {
  transition: opacity 0.3s;
}
.fade-enter,
.fade-leave-to {
  opacity: 0;
}
.overlay {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: rgba(0,0,0, 0.4);
  z-index: 998;
}
.modal {
  width: 40%;
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  box-shadow: 0px 1px 10px -5px $graphite;
  background-color: $white;
  border-radius: 4px;
  padding: 1.5em;
  z-index: 999;
}
.modal-header {
  display: flex;
  flex-wrap: nowrap;
  flex-direction: row;
  justify-content: center;
  background-color: $blue;
  padding: 0.5em;
  margin-bottom: 1em;
}
.modal-body {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
}
.modal-footer {
  display: flex;
  flex-wrap: nowrap;
  flex-direction: row;
  justify-content: flex-end;
  margin-top: 1em;
}
.modal-header-text {
  font-weight: 400;
}
</style>
