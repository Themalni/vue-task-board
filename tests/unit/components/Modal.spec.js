import { mount } from '@vue/test-utils'
import Modal from '@/components/Modal'

describe('Modal', () => {
  let wrapper

  beforeEach(() => {
    wrapper = mount(Modal)
  })

  afterEach(() => {
    wrapper.destroy()
  })

  it('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('receives text when value is provided to `title` prop', () => {
    const props = {
      title: 'Edit task'
    }
    wrapper.setProps(props)
    expect(wrapper.props('title')).toBe('Edit task')
  })
  it('dy default status element is hidden', () => {
    const modalStatus = wrapper.findAll('.modal-header-text')

    expect(wrapper.vm.status).toBe(false)
    expect(modalStatus.length).toBe(1)
  })
  it('has status element visible when prop `status` is set to true', () => {
    const props = {
      status: true
    }
    wrapper.setProps(props)

    const modalStatus = wrapper.findAll('.modal-header-text').at(1)
    expect(wrapper.props('status')).toBe(true)
    expect(modalStatus.isVisible()).toBe(true)
  })
  it('receives status text when value is provided to prop `status`', () => {
    const props = {
      status: 'Done'
    }
    wrapper.setProps(props)
    expect(wrapper.vm.status).toBe('Done')
  })
  it('by default does not receive class `w-25` when prop `small` is set to false', () => {
    expect(wrapper.vm.small).toBe(false)
    expect(wrapper.find('.modal').classes()).not.toContain('w-25')
  })
  it('receives class `w-25` when prop `small` is set to true', () => {
    const props = {
      small: true
    }
    wrapper.setProps(props)
    expect(wrapper.props('small')).toBe(true)
    expect(wrapper.find('.modal').classes()).toContain('w-25')
  })
})
