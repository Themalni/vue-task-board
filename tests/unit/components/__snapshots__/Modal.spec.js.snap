// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders correctly 1`] = `
<div class="overlay" name="fade">
  <div class="modal">
    <div class="modal-header">
      <h3 class="modal-header-text text-white"></h3>
      <!---->
    </div>
    <div class="modal-body"></div>
    <div class="modal-footer"></div>
  </div>
</div>
`;
